controllers/machinedeployment_ondelete.go

```diff
@@ -0,0 +1,147 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/controllers/mdutil"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// rolloutOnDelete implements the logic for the OnDelete MachineDeploymentStrategyType
+func (r *MachineDeploymentReconciler) rolloutOnDelete(ctx context.Context, d *clusterv1.MachineDeployment, msList []*clusterv1.MachineSet) error {
+	newMS, oldMSs, err := r.getAllMachineSetsAndSyncRevision(ctx, d, msList, true)
+	if err != nil {
+		return err
+	}
+
+	// newMS can be nil in case there is already a MachineSet associated with this deployment,
+	// but there are only either changes in annotations or MinReadySeconds. Or in other words,
+	// this can be nil if there are changes, but no replacement of existing machines is needed.
+	if newMS == nil {
+		return nil
+	}
+
+	allMSs := append(oldMSs, newMS)
+
+	// Scale up, if we can.
+	if err := r.reconcileNewMachineSetOnDelete(ctx, allMSs, newMS, d); err != nil {
+		return err
+	}
+
+	if err := r.syncDeploymentStatus(allMSs, newMS, d); err != nil {
+		return err
+	}
+
+	// Scale down, if we can.
+	if err := r.reconcileOldMachineSetsOnDelete(ctx, oldMSs, allMSs, d); err != nil {
+		return err
+	}
+
+	if err := r.syncDeploymentStatus(allMSs, newMS, d); err != nil {
+		return err
+	}
+
+	if mdutil.DeploymentComplete(d, &d.Status) {
+		if err := r.cleanupDeployment(ctx, oldMSs, d); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+//reconcileOldMachineSetsOnDelete handles reconciliation of Old MachineSets associated with the MachineDeployment in the OnDelete MachineDeploymentStrategyType
+func (r *MachineDeploymentReconciler) reconcileOldMachineSetsOnDelete(ctx context.Context, oldMSs []*clusterv1.MachineSet, allMSs []*clusterv1.MachineSet, deployment *clusterv1.MachineDeployment) error {
+	log := ctrl.LoggerFrom(ctx)
+	if deployment.Spec.Replicas == nil {
+		return errors.Errorf("spec replicas for MachineDeployment %q/%q is nil, this is unexpected",
+			deployment.Namespace, deployment.Name)
+	}
+	log.V(4).Info("Checking to see if machines have been deleted or are in the process of deleting for old machine sets")
+	totalReplicas := mdutil.GetReplicaCountForMachineSets(allMSs)
+	scaleDownAmount := totalReplicas - *deployment.Spec.Replicas
+	for _, oldMS := range oldMSs {
+		if oldMS.Spec.Replicas == nil || *oldMS.Spec.Replicas <= 0 {
+			log.V(4).Info(fmt.Sprintf("MachineSet %s fully scaled down", oldMS.Name))
+			continue
+		}
+		selectorMap, err := metav1.LabelSelectorAsMap(&oldMS.Spec.Selector)
+		if err != nil {
+			log.V(4).Error(err, "failed to convert MachineSet %q label selector to a map", oldMS.Name)
```

Comment:
nit: same as above so we log resources names consistently `ctrlclient.ObjectKeyFromObject(oldMS).String())`

Reply:
addressed

Reply:
these changes caused panics during api testing. Will not be able to port to new format and will need to keep consistent with old